plugins/woocommerce/src/Internal/Traits/RestApiCache.php

```diff
@@ -0,0 +1,405 @@
+	 * @return bool True if caching should be used, false otherwise.
+	 */
+	private function should_use_cache_for_request( WP_REST_Request $request ): bool {
+		$skip_cache   = $request->get_param( '_skip_cache' );
+		$should_cache = ! ( 'true' === $skip_cache || '1' === $skip_cache );
+
+		/**
+		 * Filter whether to enable response caching for a given REST API controller.
+		 *
+		 * @since 10.4.0
+		 *
+		 * @param bool            $enable_caching Whether to enable response caching (result of !_skip_cache evaluation).
+		 * @param object          $controller     The controller instance.
+		 * @param WP_REST_Request $request        The request object.
+		 * @return bool True to enable response caching, false to disable.
+		 */
+		return apply_filters(
+			'woocommerce_rest_api_enable_response_caching',
+			$should_cache,
+			$this,
+			$request
+		);
+	}
+
+	/**
+	 * Build the output cache entry configuration from the request and per-endpoint config.
+	 *
+	 * @param WP_REST_Request $request The request object.
+	 * @param array           $config  Raw configuration array passed to with_cache.
+	 * @return array|null Normalized cache config with keys: entity_type, cache_key. Returns null if entity type is not available.
+	 */
+	private function build_cache_config( WP_REST_Request $request, array $config ): ?array {
+		$entity_type = $config['entity_type'] ?? $this->get_default_entity_type();
+
+		if ( ! $entity_type ) {
+			wc_get_container()->get( LegacyProxy::class )->call_function(
+				'wc_doing_it_wrong',
+				__METHOD__,
+				'No entity type provided and no default entity type available. Skipping cache.',
+				'10.4.0'
+			);
+			return null;
+		}
+
+		return array(
+			'entity_type' => $entity_type,
+			'cache_key'   => $this->get_cache_key( $request, $entity_type ),
+		);
+	}
+
+	/**
+	 * Cache the response if it's successful and return it with appropriate headers.
+	 *
+	 * Only caches responses with 2xx status codes. Always adds the X-WC-Cache header
+	 * with value MISS if the response was cached, or SKIP if it was not cached.
+	 *
+	 * Supports both WP_REST_Response objects and raw data (which will be wrapped in a response object).
+	 * Error objects are returned as-is without caching.
+	 *
+	 * @param WP_REST_Request                         $request       The request object.
+	 * @param WP_REST_Response|\WP_Error|array|object $response      The response to potentially cache.
+	 * @param array                                   $cached_config Caching configuration from build_cache_config().
+	 * @return WP_REST_Response|\WP_Error The response with appropriate cache headers.
+	 */
+	private function maybe_cache_response( WP_REST_Request $request, $response, array $cached_config ) {
+		if ( is_wp_error( $response ) ) {
+			return $response;
+		}
+
+		$response = rest_ensure_response( $response );
+
+		$cached = false;
+
+		$status = $response->get_status();
+		if ( $status >= 200 && $status <= 299 ) {
```

Comment:
Are there reasons we don't want to cache other response codes?

If we're caching the response itself, I would think we'd still want to cache things like 4xx or 3xx responses.  However, that may not be the case if we're focused on etags.